src/abstractdataframe/abstractdataframe.jl

```diff
@@ -685,22 +685,32 @@ Base.vcat(df::AbstractDataFrame) = df
 Base.vcat(dfs::AbstractDataFrame...) = vcat(AbstractDataFrame[dfs...])
 
 Base.vcat(dfs::Vector{Void}) = dfs
+
 function Base.vcat{T<:AbstractDataFrame}(dfs::Vector{T})
     isempty(dfs) && return DataFrame()
-    coltyps, colnams, similars = _colinfo(dfs)
-
     res = DataFrame()
-    Nrow = sum(nrow, dfs)
-    for j in 1:length(colnams)
-        colnam = colnams[j]
-        col = similar(similars[j], coltyps[j], Nrow)
-
-        i = 1
-        for df in dfs
-            if haskey(df, colnam)
-                copy!(col, i, df[colnam])
+    nrows = sum(nrow, dfs)
+    for colnam in unique([(names(e) for e in dfs)...;])
+        k = Bool[haskey(e, colnam) for e in dfs]
+        c = vcat((e[colnam] for e in view(dfs, k))...)
+        if all(k)
+            col = c
+        else
+            col = if _isnullable(c)
+              similar(c, nrows)
```

Comment:
Sorry if I wasn't clear. The goal is to avoid an unneeded allocation. My idea was that if `isleaftype(T)`, then you can create the (empty) final array, and fill it directly with `copy!`. To be sure you can have missing values, add `NullableArray{promote_type(eltypes.(c)...)}` to the types of existing columns.

If `!isleaftype(T)`, then the only way to find out the type of that array is to create it by actually calling `vcat`. Since you need to do that, you can as well create empty `NullableArrays` of the needed length and add them to the `vcat` call so that it gives the final array directly. But honestly, I don't expect this branch to happen often in practice; we could almost raise an error.

Does that make sense?